fastestimator/network.py

```diff
@@ -176,45 +154,82 @@ def __init__(self, ops: Iterable[Union[TensorOp, Scheduler[TensorOp]]]):
         self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
 
     def load_epoch(self, mode: str, epoch: int) -> List[TensorOp]:
-        epoch_ops = super().load_epoch(mode, epoch)
+        super().load_epoch(mode, epoch)
         if self.device.type == "cuda":
             for model in self.epoch_models:
                 model.to(self.device)
-        return epoch_ops
+        update_ops = [op for op in self.epoch_ops if isinstance(op, UpdateOp)]
+        for idx, update_op in enumerate(update_ops):
+            update_op.final_update = idx == len(update_ops) - 1
 
     def unload_epoch(self):
         if self.device.type == "cuda":
             for model in self.epoch_models:
                 model.to("cpu")
 
-    def forward_step_eager(self,
-                           batch: Dict[str, Any],
-                           state: Dict[str, Any],
-                           ops: List[TensorOp],
-                           effective_outputs: List[str]) -> Dict[str, Any]:
-        prediction = {}
+    def _get_effective_batch_input(self, batch: MutableMapping[str, Any], mode: str) -> Dict[str, Any]:
+        #copy data to gpu
+        if self.device.type == "cuda":
+            new_batch = {key: batch[key].to(self.device) for key in self.effective_inputs[mode] if key in batch}
+        else:
+            new_batch = {key: batch[key] for key in self.effective_inputs[mode] if key in batch}
+        return new_batch
+
+    def run_step(self, batch: Dict[str, Any], state: Dict[str, Any]) -> Tuple[Dict[str, Any]]:
+        batch_in = self._get_effective_batch_input(batch, state["mode"])
         mode = state["mode"]
         state["tape"] = NonContext()
-        if self.device.type == "cuda":
-            for key, val in batch.items():
-                batch[key] = val.to(self.device)
+        #gpu operation
         with torch.no_grad() if mode != "train" else NonContext():
-            self._forward_batch(batch, state, ops)
-        for key in effective_outputs:
-            if key in batch:
-                value = batch[key]
-                if self.device.type == "cuda":
-                    value = value.to("cpu")
-                prediction[key] = value
-        return prediction
+            self._forward_batch(batch_in, state, self.epoch_ops)
+        #copy data to cpu
+        if self.device.type == "cuda":
+            prediction = {key: batch_in[key].to("cpu") for key in self.effective_outputs[mode] if key in batch_in}
```

Comment:
batch_in[key].detach().to("cpu")

to clear graph instead of making last backward() a retrain_graph=False